### DISCUSSION TITLE
[JS201] JS Objects 是 pass by sharing?

### DISCUSSION THREAD

**Discussion by s103071049** [Lidemy/forum#167] [General]
關於閱讀文章 : [深入探討 JavaScript 中的參數傳遞：call by value 還是 reference？](https://blog.huli.tw/2018/06/23/javascript-call-by-value-or-reference/)
很疑惑，我的理解是 js objects 應該都是 call by reference，為甚麼這樣的理解好像不太對呢 ?
![截圖](https://static.coderbridge.com/img/s103071049/bf78eae1967c4bd0b9b6a2ed8e500f21.png)

根據 C how to program with an introduction to C++ 

5.9
In many programming languages, there are two ways to pass arguments—pass-by-value
and pass-by-reference. When arguments are passed by value, a copy of the argument’s value
is made and passed to the called function. Changes to the copy do not affect an original
variable’s value in the caller. When an argument is passed by reference, the caller allows the
called function to modify the original variable’s value. 

10.5 using structures with functions
When structures or individual structure members are passed to a function, they're passed by value. Therefore, the members of a caller's structure cannot be modified by the called function.  To pass a structure by reference, pass the address of structure variable.

根據 To pass a structure by reference, pass the address of structure variable. 的理解，

```js
function add(obj) {

  obj = {
    number: obj.number + 1
  }
}

var o = {number : 10}
add(o)
console.log(o)
```
Q1 : 關於上面代碼的執行流程理解正確嗎 ?
1. 宣告 var o：0x01，其中 0x01: {number : 10}
2. 呼叫 add(o)
3. 因為 pass by reference，所以對 add function 的 obj 而言，他現在是 obj : 0x01，其中 0x01: {number : 10}
4. `  obj = { number: obj.number + 1}`，當我對 object types 用 = 賦值，他會給一塊記憶體，所以現在的 obj: 0x02，其中 0x02: {number : 11}
5. 印出 o，o :0x01 所以印出 10

Q2 : 為甚麼主流對於 js objects 的說法會是 pass by sharing ?
我的理解是以傳入的到底是值還是 memory address 進行劃分，所以會覺得 pass by sharing 好像沒有這個名詞誕生的必要。對 js 而言，object types 都是 pass by reference，可是主流的說法還是以 pass by sharing 為主 (?) 為甚麼  ? 

Q3 : 這樣的勇氣可以嗎 ?
文章最後的結論是，名詞的定義不同，多個地方都用相同的，所以造成認知不同。重點是搞清楚透過參數操作物件時，會有什麼樣的行為。所以，我可以大膽的說： 記住代碼執行產生的行為，JS objects 到底是 pass by reference 還是 pass by sharing 一點都不重要。

**Comment by allenliaooooooo** [comment]
> Q3 : 這樣的勇氣可以嗎 ?
> 文章最後的結論是，名詞的定義不同，多個地方都用相同的，所以造成認知不同。重點是搞清楚透過參數操作物件時，會有什麼樣的行為。所以，我可以大膽的說： 記住代碼執行產生的行為，JS objects 到底是 pass by reference 還是 pass by sharing 一點都不重要。

我蠻同意的～有些名詞的誕生只是為了溝通方便跟知識的傳遞，甚至是為了讓人覺得更專業（？
另外你的連結放到 prototype-chain 的了 XD

**Reply by s103071049** [reply]
對也!!!脫線的我，已修正連結，感恩。
真的，同意加一!!

**Comment by aszx87410** [comment]
> Q1 : 關於上面代碼的執行流程理解正確嗎 ?

正確

> Q2: 為甚麼主流對於 js objects 的說法會是 pass by sharing ?

可以回去複習一下文章裡面最下面的圖，如果要是 pass by reference，那你在 function 裡面對 object 重新賦值，function 外面的 object 也會一起被改掉，但是 JS 不是這樣，所以不能稱之為是 pass by reference

> Q3 : 這樣的勇氣可以嗎 ?

從「文章最後的結論是，名詞的定義不同，多個地方都用相同的，所以造成認知不同。重點是搞清楚透過參數操作物件時，會有什麼樣的行為」，可以推論出：「記住代碼執行產生的行為」是重要的，但不能推論出「JS objects 到底是 pass by reference 還是 pass by sharing 一點都不重要」

事實上也並不是一點都不重要啦，就多少知道一下這兩個名詞在幹嘛還是有需要的，之後跟同事討論溝通什麼的也比較方便。但如果我同事說某個 bug 是因為 pass by reference 引起的，我也不會糾正他說：「這叫 pass by sharing」，因為我知道他想表達的是什麼意思

**Reply by s103071049** [reply]
豁然開朗的感覺，謝謝老師的回答!!
